cellmlmanip/_singularity_fixes.py

```diff
@@ -0,0 +1,252 @@
+        return m is not None and Z in m and Z != 0 and (sp == m[SP_wildcard]
+                                                        or (isinstance(sp, Float)
+                                                            and isinstance(m[SP_wildcard], Float)
+                                                            and isclose(m[SP_wildcard], sp)))
+
+    def float_dummies(expr):
+        """Turns floats back into Quantity dummies to be in line with the rest of the sympy equations"""
+        if expr is None:
+            return None
+        return expr.xreplace({f: cellmlmanip.Quantity(f, 'dimensionless') for f in expr.atoms(Float)})
+
+    # the denominator is all args where a **-1
+    numerator = tuple(a for a in expr.args if not isinstance(a, Pow) or a.args[1] != -1.0)
+    denominator = tuple(a.args[0] for a in expr.args if isinstance(a, Pow) and a.args[1] == -1.0)
+
+    # Not a devision or does not have exp
+    if len(denominator) == 0 or len(numerator) == 0 or not expr.has(exp_function):
+        return None, None, None
+
+    (Vmin, Vmax, sp) = None, None, None
+    # U might be on top, try numerator / denominator and denominator / numerator
+    for num, denom in ((numerator, denominator), (denominator, numerator)):
+        found_on_top = False
+        for d in denom:  # Check arguments in denominator (or numerator)
+            if d.has(exp_function):
+                find_U = d.match(exp_function(U_wildcard) * -Z + 1.0)  # look for exp(U) * -Z + 1.0
+                if not check_bottom_match(find_U):
+                    find_U = d.match(exp_function(U_wildcard) * Z - 1.0)  # look for exp(U) * Z - 1.0
+                if check_bottom_match(find_U):
+                    # We found a match, since exp(U) * Z == exp(U + log(Z)) we can bring Z into the u expression
+                    u = (find_U[U_wildcard] + log(find_U[Z]))
+                    try:
+                        # Find the singularity point by solving u for V==0, excluding irratinal results
+                        sp = tuple(filter(lambda s: not s.has(I), solveset(u, V)))
+                        if sp:
+                            # we found a singularity point, now find Vmin, Vmax
+                            # by solving for U_offset either side of U
+                            find_v_low = solveset(u + U_offset, V)
+                            find_v_up = solveset(u - U_offset, V)
+                            find_v_low = tuple(find_v_low)
+                            find_v_up = tuple(find_v_up)
+                            assert find_v_low and find_v_up and len(find_v_low) == len(find_v_up) == len(sp) == 1, \
+                                'Expecting exactly 1 singularity point '
+                            (Vmin, Vmax, sp) = (find_v_low[0], find_v_up[0], sp[0])
+                    except TypeError:
+                        pass  # Result could be 'ConditionSet' which is not iterable and not Real
+
+                if Vmin is not None:  # check top
```

Comment:
Is this just "if there wasn't a type error" ? Some way to write this bit slightly nicer? Maybe a `continue` in the "except TypeError" clause?
(And then can any of those lines throw a typeerror, or can we make it more specific?

Reply:
Moved the solving into it's own function.
`_solve_tuple`